## 1. Base Python

### 1.1 Computing

1. How many seconds are there in year?

In [1]:
secInYear = 60 * 60 * 24 *365
print(secInYear)

31536000


3153600 seconds in a year

2. How many seonds is a typical human lifetime? Use the previously computed seconds in year.

In [2]:
secInLifetime = secInYear * 74.5
print(secInLifetime)

2349432000.0


About 2349432000 seconds in a typical 74.5 years lifespan

3. What is your age in seconds? Compute this based on your age.

In [3]:
secInMyAge = secInYear * 20
print(secInMyAge)

630720000


My age in seconds is 630720000 seconds.

4. Create a logical variable that is true if you are more than 700M seconds old. Use logical operators, not if/else!

In [4]:
isMoreThanSevenMil = secInMyAge > 700000000
print(isMoreThanSevenMil)

False


5. Compute the following mathematical formulas

In [5]:
p = 0.4
firstQ = p * (1 - p)
print(firstQ)
alpha = 3
x0 = 2
x = 2.5
secondQ = (alpha / x0) * ((1 + (x / x0)) ** (-alpha - 1))
print(secondQ)

0.24
0.058527663465935076


### 1.2 Lists

1. Create a list ‘movies‘ that contains the names of at least six movies you like

In [6]:
favMovies = ["Inception", "Spiderman", "Fast Five", "Star Wars: A New Hope", "Kung Fu Panda", "Jurassic Park"]
print(favMovies)

['Inception', 'Spiderman', 'Fast Five', 'Star Wars: A New Hope', 'Kung Fu Panda', 'Jurassic Park']


2. Using indexing and slicing, Create a list of three first movies in the list

In [7]:
firstThreeMovies = favMovies[:3]
print(firstThreeMovies)

['Inception', 'Spiderman', 'Fast Five']


3. Use slicing and list concatenation to create a list of the first two and the last two movies

In [8]:
firstTwoLastTwo = favMovies[:2] + favMovies[-2:]
print(firstTwoLastTwo)

['Inception', 'Spiderman', 'Kung Fu Panda', 'Jurassic Park']


4. Use slicing to list the movies in the opposite order

In [9]:
reverseMovieList = favMovies[::-1]
print(reverseMovieList)

['Jurassic Park', 'Kung Fu Panda', 'Star Wars: A New Hope', 'Fast Five', 'Spiderman', 'Inception']


### 1.3 Loops

1. Create a list ‘numbers‘ that is the numbers 70 through 79 in the following manner: create an empty list and add numbers to it in a loop (do not use list comprehension or ‘list‘ function here!)

In [10]:
numList = []
for x in range(10):
    numList.append(70 + x)
print(numList)

[70, 71, 72, 73, 74, 75, 76, 77, 78, 79]


2. Use a loop to compute the mean value of the list: add the values to their sum in a loop, and at the end divide the sum by the number of the values. (Do not use mean function!)

In [11]:
total = 0
for x in numList:
    total += x
avg = total/len(numList)
print(avg)

74.5


The average from 70-79 is 74.5

3. Use loop to compute sum of all integers from 1 till 100

In [12]:
total = 0
for x in range(100):
    total += (1 + x)
print(total)

5050


The sum from 1 - 100 is 5050.

4. Assign people to seats. Consider names Adam, Ashin, Inukai, Tanaka, and Ikki; and seats 33, 12, 45, 2 and 17. Print seat assignments by assigning each name to the corresponding seat.

In [13]:
name = ["Adam", "Ashin", "Inukai", "Tanaka", "Ikki"]
seats = [33, 12, 45, 2, 17]
for x in range(len(name)):
    print(name[x] + ": " + str(seats[x]))

Adam: 33
Ashin: 12
Inukai: 45
Tanaka: 2
Ikki: 17


### 1.4 Dicts

1. Create a word frequency table using dicts that takes a looong string (feel free to copy-paste some text here), splits it into individual words, and counts the number of occurrencies of each word (and prints the result). Let’s ignore punctuation. Your code should run over individual words and increase the counter for that word, stored in a list. It has to check if a word exists in the dictionary, and if not, take an appropriate action.

In [14]:
testString = "He likes a dog who likes to eat a hot dog"

def wordTable(sentence):
    wordList = sentence.lower().split()
    wordDict = {}
    for w in wordList:
        if w not in wordDict.keys():
            wordDict[w] = 1
        else:
            wordDict[w] = wordDict[w] + 1
    return wordDict

print(wordTable(testString))

{'he': 1, 'likes': 2, 'a': 2, 'dog': 2, 'who': 1, 'to': 1, 'eat': 1, 'hot': 1}


### 1.5 Functions and data transformation

1. Write a function that takes in time in the numeric form (i.e. not a string) of HHMM (hours- minutes), and returns it in the numeric form of HH.HH (hours + fractions of hours). For instance, 1015 → 10.25 (10 hrs 15 mins → 10.25 hrs). Demonstrate it works using values 1015 and 345

In [15]:
def hourTransformation(time):
    minutes = time % 100 / 60
    hours = time // 100
    return hours + minutes

print(hourTransformation(1015))
print(hourTransformation(345))

10.25
3.75


10.25 hr
3.75 hr

## 2. Basic data frames

### 2.1 Setup

1. Load the data

In [16]:
import numpy as np
import pandas as pd

In [17]:
nycflights = pd.read_csv('PS 1/nycflights13.csv.bz2', sep="\t")

2a. How many variables (columns) is there in the data? Ensure you know the variables in the data. Keep the documentation nearby.

In [18]:
print(nycflights.shape)

(336776, 19)


19 variables

2b. How many rows of data is there?

336776 rows of data

2c. print the first few lines of data. Does it look reasonable?

In [19]:
print(nycflights.sample(3))

        year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
189814  2013      4   27     557.0             600       -3.0     851.0   
287283  2013      8    8    1743.0            1709       34.0     111.0   
271132  2013      7   22       NaN            2000        NaN       NaN   

        sched_arr_time  arr_delay carrier  flight tailnum origin dest  \
189814             854       -3.0      B6     501  N639JB    JFK  FLL   
287283            2035        NaN      UA    1284  N37456    EWR  SFO   
271132            2136        NaN      UA    1128     NaN    LGA  ORD   

        air_time  distance  hour  minute             time_hour  
189814     145.0      1069     6       0  2013-04-27T10:00:00Z  
287283       NaN      2565    17       9  2013-08-08T21:00:00Z  
271132       NaN       733    20       0  2013-07-23T00:00:00Z  


The data set looks reasonable, although I did notice some NaN values in other rows.

### 2.2 Basic Data Exploration

1. How many NYC airports are included in this data? Which airports are these?

In [20]:
uniqueAirports = nycflights.origin.unique()
print(uniqueAirports)
print(len(uniqueAirports))

['EWR' 'LGA' 'JFK']
3


There are three NYC airports EWR, LGA, and JFK

2. Into how many airports did the airlines fly from NYC in 2013?

In [21]:
uniqueDestAirports = nycflights.dest.unique()
print(len(uniqueDestAirports))

105


There are 105 airports that NYC flights fly into

3. How many flights were there from NYC to Seattle (airport code SEA)?

In [22]:
flightsToSea = nycflights[nycflights.dest == "SEA"].copy()
print(len(flightsToSea.index))

3923


There were 3923 flights into Seattle.

4. Were there any flights from NYC to Spokane (GEG)?

In [23]:
flightsToGEG = nycflights[nycflights.dest == "GEG"].copy()
print(len(flightsToGEG.index))

0


There were 0 flights from NYC to Spokane.

### 2.3 Compute new variables

1. Create new column for the flight data–flight duration in hours. Use variables “arr time” and “dep time”, not “air time”.

In [24]:
nycflightsWithDuration = nycflights.copy()
nycflightsWithDuration["flightDuration"] = hourTransformation(nycflightsWithDuration["arr_time"] - nycflightsWithDuration["dep_time"])

2. What is the minimum value of the duration you computed? Explain why do you get negative values!

In [25]:
minFlightDuration = nycflightsWithDuration[nycflightsWithDuration.flightDuration == nycflightsWithDuration.flightDuration.min()]
print(minFlightDuration.flightDuration)

266385   -23.1
Name: flightDuration, dtype: float64


The minimum value for flight duration is -23.1. We get negative values if the arrive time is AM of the next day since AM time values would be less than PM time values.

### 2.4 Flight delays

1. Create a new data frame, a subset of the flights data that only contains two variables: departure delay and arrival delay.

In [26]:
flightDelays = nycflights[["dep_delay", "arr_delay"]].copy()

2. Print a small random sample of the delay data frame.

In [27]:
print(flightDelays.sample(3))

        dep_delay  arr_delay
304331       -5.0      -21.0
178985       -7.0        3.0
334441       -1.0      -28.0


3. What is the data type of the flight delay value?

In [28]:
print(type(flightDelays.iloc[0].dep_delay))
print(type(flightDelays.iloc[0].arr_delay))

<class 'numpy.float64'>
<class 'numpy.float64'>


The datatype for the two flight delay values is float64.

4. Filter to keep the “slow flights”: cases where the plane leaves early but arrives late. How many cases do you have?

In [29]:
print(len(flightDelays.loc[(flightDelays["arr_delay"] > 0) & (flightDelays["dep_delay"] < 0)].index))

35648


There are 35648 cases of slow flights.

5. What is the longest “extra flight time”, i.e. difference between arrival delay and departure delay?

In [30]:
print(max(hourTransformation(flightDelays.arr_delay - flightDelays.dep_delay)))

2.6


The longest extra flight time is 2.6 hours.

6. How many of flights were delayed (positive arrival delay)?

In [31]:
print(len(flightDelays.loc[flightDelays["arr_delay"] > 0].index))

133004


There are 133004 flights were delayed

7. What percentage of flights were delayed by more than 15 minutes (at arrival)?

In [32]:
numFlights15Delay = (len(flightDelays.loc[flightDelays["arr_delay"] > 15].index))
percent15Delay = numFlights15Delay/len(nycflights.index)
print(percent15Delay)

0.2305092999501152


About 23.05% of flights were delayed by more than 15 minutes.

## How much time did you spend?

1.5 hours